### PR TITLE
Update domain rules

### PR DESCRIPTION
**Pull request by nahteb**
Split this into 2 rules, one to check that the domain is the correct length and
one to check it contains the correct characters. This makes it easier to see
why this causes an error if the domain doesn't follow one or both of these
rules.

We also need to include the scheme in the regex, which should be `https` unless
the url is `http://localhost`.

### CODE REVIEW

File: spectral-ruleset-govuk-public/ruleset.yaml [hunk 1]
```diff
@@ -74,11 +74,19 @@ rules:
...
+  domain-name-length:
     severity: error
     given: "$.servers[*].url"
-    description: "The domain name must be between 3 and 63 characters long, and contain only alphanumeric characters (0-9 and a-z) and the ‘-’ (dash) symbol. It should also use dashes between words to make your domain easy to read - for example, vehicle-registration-number.api.gov.uk."
+    description: "The domain name must be between 3 and 63 characters long."
     then:
       function: pattern
       functionOptions:
-        match: "^[a-zA-Z0-9][a-zA-Z0-9-]{1,61}[a-zA-Z0-9](?:\\.[a-zA-Z]{2,})+$"
+        match: "(http:\/\/localhost$)|(https:\/\/[^\\.]{3,63}(?:\\.[a-zA-Z]{2,})+$)"
```
[the remark this answers is not in the document]
Comment (RobNicholsGDS) [reply spectral-ruleset-govuk-public/ruleset.yaml:84]: I wonder if this is a better solution:
```
(http:\/\/localhost$)|(https:\/\/([^\.]{2,63}\.)+[^\.]{2,63}$)
```
That's:
| regex element | description |
|---------------|------------|
|http:\/\/localhost$| `http://localhost` |
| `()\|()` | or |
|`https:\/\/`| `https//`|
|`([^\.]{2,63}\.)+`| Any number of sets of 2-63 non-dot characters followed by a dot |
|`[^\.]{2,63}`| A single 2-63 set of non-dot characters |
| `$` | end of string |

File: spectral-ruleset-govuk-public/ruleset.yaml [hunk 2]
```diff
@@ -74,11 +74,19 @@ rules:
...
+  domain-name-characters:
+    severity: error
+    given: "$.servers[*].url"
+    description: "The domain name must contain only alphanumeric characters (0-9 and a-z) and the ‘-’ (dash) symbol."
+    then:
+      function: pattern
+      functionOptions:
+        match: "(http:\/\/localhost$)|(https:\/\/[a-zA-Z0-9][a-zA-Z0-9-]{1,}[a-zA-Z0-9](?:\\.[a-zA-Z]{2,})+$)"
```
Comment (RobNicholsGDS) [inline spectral-ruleset-govuk-public/ruleset.yaml:92]: Again, delimit the string with single quotes.

This regex only tests that the first element of the domain name contains letters, numbers and hyphens. Any element after that can only have letters. So:

| url       | expected | actual |
| ------ | --------- | -------|
| `http://localhost` | pass | pass |
|`https://google.com`| pass | pass |
|`https://this-that.google.com`|pass|pass|
|`https://this_that.google.com`|fail|fail|
|`https://google.this-that.com`|pass|fail|
|`https://123.google.com`|pass|pass|
|`https://google.123.com`|pass|fail|

I'd suggest:
```
(http:\/\/localhost$)|(https:\/\/([a-zA-Z0-9][a-zA-Z0-9\-]*[a-zA-Z0-9]\.)+([a-zA-Z]{2,})$)
```

**Review by RobNicholsGDS — CHANGES_REQUESTED** [review]
The regular expression will be easier to test outside the app if they are delimited by single quotes. That's minor.

I've requested changes because both regular expressions only test the first element of the domain name.

**Comment by nahteb** [comment]
@RobNicholsGDS thanks, very good point about it only checking the first part! I've updated the regexes to use your suggestions (the only difference is on the length one I've changed the first 2 to a 3 so that domains shorter than 3 characters still fail) and added some more tests.

**Review by RobNicholsGDS — APPROVED** [review]
Excellent. Everything appears to be working as I would expect.

**Pull request merged**